File: packages/android/app/src/main/java/com/truthpulse/data/SynonymTable.kt
package com.truthpulse.data

import android.content.Context
import kotlinx.serialization.json.*

/**
 * Loads synonyms from assets/synonyms.json and expands query tokens at search time.
 */
class SynonymTable(context: Context) {

    private val table: Map<String, List<String>>

    init {
        table = loadFromAssets(context)
    }

    /**
     * Expand a list of query tokens using synonyms.
     * For each token that matches a synonym key, the expansions are appended.
     * Returns the original tokens plus any expanded terms.
     */
    fun expandTokens(tokens: List<String>): List<String> {
        val expanded = tokens.toMutableList()
        for (token in tokens) {
            val synonyms = table[token] ?: continue
            for (synonym in synonyms) {
                val synonymTokens = synonym.normalizedSearchText().searchTokens()
                for (st in synonymTokens) {
                    if (st !in expanded) {
                        expanded.add(st)
                    }
                }
            }
        }
        return expanded
    }

    private fun loadFromAssets(context: Context): Map<String, List<String>> {
        return try {
            val text = context.assets.open("synonyms.json").bufferedReader().readText()
            val root = Json.parseToJsonElement(text).jsonObject
            val synonymsObj = root["synonyms"]?.jsonObject ?: return emptyMap()

            val result = mutableMapOf<String, List<String>>()
            for ((key, value) in synonymsObj) {
                val values = (value as? JsonArray)?.mapNotNull { element ->
                    (element as? JsonPrimitive)?.contentOrNull
                } ?: continue
                result[key.lowercase()] = values
            }
            result
        } catch (_: Exception) {
            emptyMap()
        }
    }
}

// --- Search text normalization extensions ---

/**
 * Normalize text for search: lowercase, remove diacritics, collapse whitespace.
 */
fun String.normalizedSearchText(): String {
    return java.text.Normalizer.normalize(this, java.text.Normalizer.Form.NFD)
        .replace(Regex("[\\p{InCombiningDiacriticalMarks}]"), "")
        .lowercase()
        .replace(Regex("\\s+"), " ")
        .trim()
}

/**
 * Split normalized text into searchable tokens.
 */
fun String.searchTokens(): List<String> {
    return split(Regex("[^a-z0-9]+")).filter { it.isNotEmpty() }
}

File: packages/mac/Sources/TruthPulseCore/Services/SynonymTable.swift
import Foundation

/// Loads synonyms from the shared synonyms.json and expands query tokens at search time.
public struct SynonymTable: Sendable {
    private let table: [String: [String]]

    public init() {
        self.table = SynonymTable.load()
    }

    /// Expand a list of query tokens using synonyms.
    /// For each token, if it matches a synonym key, the expansions are appended.
    /// Returns the original tokens plus any expanded terms.
    public func expandTokens(_ tokens: [String]) -> [String] {
        var expanded = tokens
        for token in tokens {
            if let synonyms = table[token] {
                for synonym in synonyms {
                    let synonymTokens = synonym.normalizedSearchText.searchTokens
                    for st in synonymTokens where !expanded.contains(st) {
                        expanded.append(st)
                    }
                }
            }
        }
        return expanded
    }

    /// Expand a normalized query string into multiple query variants.
    /// Returns the original query plus queries with synonym expansions.
    public func expandQuery(_ normalizedQuery: String) -> [String] {
        let tokens = normalizedQuery.searchTokens
        let expandedTokens = expandTokens(tokens)
        let newTokens = expandedTokens.filter { !tokens.contains($0) }
        if newTokens.isEmpty {
            return [normalizedQuery]
        }
        return [normalizedQuery, newTokens.joined(separator: " ")]
    }

    private static func load() -> [String: [String]] {
        guard let url = Bundle.main.url(forResource: "synonyms", withExtension: "json")
                ?? bundleFallbackURL() else {
            return [:]
        }

        do {
            let data = try Data(contentsOf: url)
            let wrapper = try JSONDecoder().decode(SynonymFile.self, from: data)
            // Normalize all keys to lowercase
            var normalized: [String: [String]] = [:]
            for (key, values) in wrapper.synonyms {
                normalized[key.lowercased()] = values
            }
            return normalized
        } catch {
            return [:]
        }
    }

    /// Fallback: look for synonyms.json relative to the executable (for dev builds).
    private static func bundleFallbackURL() -> URL? {
        let executableURL = Bundle.main.executableURL ?? URL(fileURLWithPath: CommandLine.arguments[0])
        // Walk up from .build/debug/TruthPulse to find packages/shared/synonyms.json
        var dir = executableURL.deletingLastPathComponent()
        for _ in 0..<6 {
            let candidate = dir.appendingPathComponent("packages/shared/synonyms.json")
            if FileManager.default.fileExists(atPath: candidate.path) {
                return candidate
            }
            dir = dir.deletingLastPathComponent()
        }
        return nil
    }
}

private struct SynonymFile: Decodable {
    let synonyms: [String: [String]]
}
